tools/scripts/rebuildAnimationLibraryManifest.rb

```diff
@@ -109,6 +107,83 @@ def rebuild_animation_library_manifest
     EOS
   end
 
+  #
+  # Main entry point for library download feature
+  # See end of file for command line options
+  #
+  def download_entire_animation_library
+    # Connect to S3 and get a listing of all objects in the animation library bucket
+    bucket = Aws::S3::Bucket.new(DEFAULT_S3_BUCKET)
+    animation_objects = get_animation_objects(bucket)
+    info "Found #{animation_objects.size} animations."
+
+    info "Downloading library..."
+
+    download_progress_bar = ProgressBar.create(total: animation_objects.size) unless @options[:verbose] || @options[:quiet]
+
+    # Parallelize downloads
+    Parallel.map(animation_objects.keys, finish: lambda do |_, _, result|
+      # This lambda runs synchronously after each entry is done
+      if result.is_a? String
+        @warnings.push result
+      end
+      download_progress_bar.increment unless download_progress_bar.nil?
```

Comment:
This actually runs on the main thread - it's part of the `finish` lambda that runs on the main thread after each work item completes.